src/iptvsimple/CatchupController.cpp

```diff
@@ -0,0 +1,436 @@
+    m_timeshiftBufferStartTime = timeNow - timeshiftBufferDuration;
+    m_catchupStartTime = m_timeshiftBufferStartTime;
+    m_catchupEndTime = timeNow;
+    m_timeshiftBufferOffset = timeshiftBufferDuration - programmeOffset;
+
+    m_resetCatchupState = false;
+
+    // TODO: Need a method of updating an inputstream if already running such as web call to stream etc.
+    // this will avoid inputstream restarts which are expensive, may be better placed in client.cpp
+    // this also mean knowing when a stream has stopped
+    SetCatchupInputStreamProperties(true, channel, catchupProperties);
+  }
+  else
+  {
+    UpdateProgrammeFrom(epgTag, channel.GetTvgShift());
+    m_catchupStartTime = epgTag.startTime;
+    m_catchupEndTime = epgTag.endTime;
+
+    m_timeshiftBufferStartTime = 0;
+    m_timeshiftBufferOffset = 0;
+
+    m_fromEpgTag = true;
+  }
+}
+
+void CatchupController::ProcessEPGTagForVideoPlayback(const EPG_TAG& epgTag, Channel& channel, std::map<std::string, std::string>& catchupProperties) // TODO: should be a const channel, create a StreamManager instead
+{
+  TestAndStoreStreamType(channel); // TODO: StreamManager
+
+  if (m_controlsLiveStream)
+  {
+    if (m_resetCatchupState)
+    {
+      UpdateProgrammeFrom(epgTag, channel.GetTvgShift());
+      m_catchupStartTime = epgTag.startTime;
+      m_catchupEndTime = epgTag.endTime;
+
+      const time_t beginBuffer = Settings::GetInstance().GetCatchupWatchEpgBeginBufferSecs();
+      const time_t endBuffer = Settings::GetInstance().GetCatchupWatchEpgEndBufferSecs();
+      m_timeshiftBufferStartTime = m_catchupStartTime - beginBuffer;
+      m_catchupStartTime = m_timeshiftBufferStartTime;
+      m_catchupEndTime += endBuffer;
+      m_timeshiftBufferOffset = beginBuffer;
+
+      m_resetCatchupState = false;
+    }
+
+    // TODO: Need a method of updating an inputstream if already running such as web call to stream etc.
+    // this will avoid inputstream restarts which are expensive, may be better placed in client.cpp
+    // this also mean knowing when a stream has stopped
+    SetCatchupInputStreamProperties(false, channel, catchupProperties);
+  }
+  else
+  {
+    UpdateProgrammeFrom(epgTag, channel.GetTvgShift());
+    m_catchupStartTime = epgTag.startTime;
+    m_catchupEndTime = epgTag.endTime;
+
+    m_timeshiftBufferStartTime = 0;
+    m_timeshiftBufferOffset = 0;
+    m_catchupStartTime = m_catchupStartTime - Settings::GetInstance().GetCatchupWatchEpgBeginBufferSecs();
+    m_catchupEndTime += Settings::GetInstance().GetCatchupWatchEpgEndBufferSecs();
+  }
+
+  if (m_catchupStartTime > 0)
+    m_playbackIsVideo = true;
+}
+
+void CatchupController::SetCatchupInputStreamProperties(bool playbackAsLive, const Channel& channel, std::map<std::string, std::string>& catchupProperties)
+{
+  catchupProperties.insert({"epgplaybackaslive", playbackAsLive ? "true" : "false"});
+
+  catchupProperties.insert({"inputstream.ffmpegdirect.is_realtime_stream", "true"});
```

Comment:
How does the addon behave if inputstream.ffmpegdirect is not installed? Would be cool if the user would at least get a meaningful error message with instructions how to "fix" this.

Reply:
Hopefully to fixed by Alwin.